etuutt_bot/services/sync.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+from datetime import time
+from typing import TYPE_CHECKING
+from zoneinfo import ZoneInfoNotFoundError
+
+import aiohttp
+from discord.ext import tasks
+from pydantic import Field, ValidationError
+
+from etuutt_bot.routes.role import ApiUserSchema
+from etuutt_bot.services.user import UserService
+
+if TYPE_CHECKING:
+    from etuutt_bot.bot import EtuUTTBot
+
+
+class ApiUserSchemaDiscord(ApiUserSchema):
+    discord_tag: str = Field(alias="discordTag")
+
+
+class SyncService:
+    """Service de gestion de la synchronisation des membres du serveur."""
+
+    def __init__(self, bot: EtuUTTBot):
+        self._bot = bot
+
+        try:
+            from zoneinfo import ZoneInfo
+
+            self._timezone = ZoneInfo(bot.settings.bot.tz)
+        # On utilise la timezone UTC en cas de problème
+        except ZoneInfoNotFoundError:
+            from datetime import timezone
+
+            self._timezone = timezone.utc
+
+        @tasks.loop(time=time(0, tzinfo=self._timezone))
+        async def daily_etu_sync():
+            self._bot.logger.info("Automatic synchronisation started")
+            await self._full_sync()
+            self._bot.logger.info("Automatic synchronisation ended")
+
+        if bot.settings.guild.etu_sync:
+            daily_etu_sync.start()
+
+        self._daily_etu_sync = daily_etu_sync
+        self.is_running = daily_etu_sync.is_running
+
+    async def _full_sync(self):
+        api_settings = self._bot.settings.etu_api
+        auth = aiohttp.BasicAuth(
+            str(api_settings.client_id), api_settings.client_secret.get_secret_value()
+        )
+        data = {"grant_type": "client_credentials"}
+        async with self._bot.session.post(
+            f"{api_settings.url}/oauth/token", auth=auth, data=data
+        ) as response:
+            if response.status != 200:
+                self._bot.logger.error(
+                    "Failed to get access to API, could not perform automatic synchronisation"
+                )
+                return
+            try:
+                token = (await response.json()).get("access_token")
+            except KeyError:
+                self._bot.logger.error(
+                    "Failed to get access to API, could not perform automatic synchronisation"
+                )
+                return
+
+        user_service = UserService(self._bot)
+        api_url = str(api_settings.url).replace("/api", "")
```

Comment:
```suggestion
        api_url = str(api_settings.url).removesuffix("/api")
```
Ca devrait donner presque toujours donner le même résultat, mais avec la nuance que c'est à la fin qu'il faut retirer ça.

Reply:
On devrait pouvoir enlever ça ? et modifier le next page pour retirer le /api ?
je ne sais pas trop comment fonctionne l'api, donc je me pose la question

Reply:
J'ai l'impression qu'en fait la réponse contient le lien vers la page suivante (en fin de boucle, le `next_page = resp.get("pagination").get("next")`) et que ce lien contient le préfixe "/api", qui peut potentiellement être redondant avec l'url de base qu'on utilise.

Reply:
Oui c'est ça, j'ai noté le lien avec le /api pcq on fait que toucher l'api dans les requêtes sauf que le next contient le chemin sur le domain donc /api/public/...
C'est donc le remplacement que je voulais.